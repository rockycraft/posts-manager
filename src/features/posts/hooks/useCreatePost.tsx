import type { State } from "@/types/State";
import { useCallback, useState } from "react";
import { useDispatch } from "react-redux";
import { createPost } from "../slices/postsSlice";
import { postPost } from "../services/postPost";
import type { Post } from "@/types/Post";

export function useCreatePost() {
	const [state, setState] = useState<State>("idle");
	const dispatch = useDispatch();

	const create = useCallback(
		async (post: Omit<Post, "id">) => {
			try {
				setState("loading");
				const response = await postPost(post);
				dispatch(createPost(response));
				setState("success");
			} catch (error) {
				setState("error");
			}
		},
		[dispatch],
	);

	return { create, state };
}
