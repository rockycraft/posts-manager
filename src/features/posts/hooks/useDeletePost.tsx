import type { State } from "@/types/State";
import { useState } from "react";
import { deletePost as deletePostService } from "../services/deletePost";
import { useDispatch } from "react-redux";
import { removePost } from "../slices/postsSlice";
import { Post } from "@/types/Post";

export function useDeletePost() {
	const [state, setState] = useState<State>("idle");
	const dispatch = useDispatch();

	async function deletePost(id: Post["id"]) {
		try {
			setState("loading");
			//@ts-ignore
			const response = await deletePostService(id);
			dispatch(removePost(response.id));
			setState("success");
		} catch (error) {
			console.error(error);
			setState("error");
		}
	}

	return { deletePost, state };
}
