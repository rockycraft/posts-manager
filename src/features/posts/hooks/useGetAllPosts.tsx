import type { RootState } from "@/store";
import { useCallback, useEffect, useRef, useState } from "react";
import { shallowEqual, useDispatch, useSelector } from "react-redux";
import { getPosts } from "../services/getPosts";
import { addPosts } from "../slices/postsSlice";
import type { State } from "@/types/State";

export function useGetAllPosts() {
	const [state, setState] = useState<State>("idle");
	const initialized = useRef(false);
	const posts = useSelector(
		(state: RootState) => state.posts.posts,
		shallowEqual,
	);
	const dispatch = useDispatch();

	const getAllPosts = useCallback(async () => {
		try {
			setState("loading");
			const response = await getPosts();
			dispatch(addPosts(response));
			setState("success");
		} catch (error) {
			setState("error");
			console.error(error);
		}
	}, [dispatch]);

	useEffect(() => {
		if (!initialized.current) {
			initialized.current = true;

			getAllPosts();
		}
	}, [getAllPosts]);

	return { data: posts, state, getAllPosts };
}
