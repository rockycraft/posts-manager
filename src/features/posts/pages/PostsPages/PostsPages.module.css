.container {
	height: 100vh;
	width: 100%;
	padding: 32px 64px;
}
