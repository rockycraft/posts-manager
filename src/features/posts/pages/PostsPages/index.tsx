import { memo, useState } from "react";
import Table, { type ColumnFilters } from "@/features/posts/components/Table";
import { useGetAllPosts } from "../../hooks/useGetAllPosts";
import styles from "./PostsPages.module.css";
import { Show, Stack } from "@chakra-ui/react";
import NewPostForm from "../../components/NewPostForm";
import SearchPostForm from "../../components/SearchPostForm";
import NoDataMessage from "../../components/NoDataMessage";

function PostsPages() {
	const { data } = useGetAllPosts();
	const [columnFilters, setColumnFilters] = useState<ColumnFilters>([]);

	return (
		<div className={styles.container}>
			<Stack gap="5">
				<Show when={data.length > 0} fallback={<NoDataMessage />}>
					<SearchPostForm onSearch={setColumnFilters} />
					<Table data={data} columnFilters={columnFilters} />
				</Show>
				<NewPostForm />
			</Stack>
		</div>
	);
}

export default memo(PostsPages);
