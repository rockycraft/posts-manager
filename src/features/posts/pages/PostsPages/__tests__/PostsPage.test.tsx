import "@testing-library/jest-dom";
import { render, screen } from "@testing-library/react";
import { describe, it, expect, vi, beforeEach } from "vitest";
import PostsPages from "../index";
import { useGetAllPosts } from "../../../hooks/useGetAllPosts";
import { Provider as ChakraProvider } from "@/components/ui/provider";
import { Provider } from "react-redux";
import { store } from "@/store";

Object.defineProperty(window, "matchMedia", {
	writable: true,
	value: vi.fn().mockImplementation((query) => ({
		matches: false,
		media: query,
		onchange: null,
		addListener: vi.fn(), // deprecated
		removeListener: vi.fn(), // deprecated
		addEventListener: vi.fn(),
		removeEventListener: vi.fn(),
		dispatchEvent: vi.fn(),
	})),
});

vi.mock("../../../hooks/useGetAllPosts");

describe("PostsPages component", () => {
	const mockData = [
		{ id: "1", name: "Post 1", description: "Description 1" },
		{ id: "2", name: "Post 2", description: "Description 2" },
	];

	beforeEach(() => {
		vi.clearAllMocks();
		// @ts-ignore
		useGetAllPosts.mockReturnValue({ data: mockData });
	});

	it("should render the PostsPages component with data", () => {
		render(
			<Provider store={store}>
				<ChakraProvider>
					<PostsPages />
				</ChakraProvider>
			</Provider>,
		);

		expect(screen.getByText("Post 1")).toBeInTheDocument();
		expect(screen.getByText("Post 2")).toBeInTheDocument();
	});

	it("should render NoDataMessage when there is no data", () => {
		// @ts-ignore
		useGetAllPosts.mockReturnValue({ data: [] });

		render(
			<Provider store={store}>
				<ChakraProvider>
					<PostsPages />
				</ChakraProvider>
			</Provider>,
		);

		expect(screen.getByText("No hay posts para mostrar")).toBeInTheDocument();
	});

	it("should match snapshot", () => {
		const { asFragment } = render(
			<Provider store={store}>
				<ChakraProvider>
					<PostsPages />
				</ChakraProvider>
			</Provider>,
		);
		expect(asFragment()).toMatchSnapshot();
	});
});
