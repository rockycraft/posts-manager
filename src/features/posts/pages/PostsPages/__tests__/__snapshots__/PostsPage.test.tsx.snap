// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PostsPages component > should match snapshot 1`] = `
<DocumentFragment>
  <script
    nonce=""
  >
    ((e,i,s,u,m,a,l,h)=&gt;{let d=document.documentElement,w=["light","dark"];function p(n){(Array.isArray(e)?e:[e]).forEach(y=&gt;{let k=y==="class",S=k&&a?m.map(f=&gt;a[f]||f):m;k?(d.classList.remove(...S),d.classList.add(a&&a[n]?a[n]:n)):d.setAttribute(y,n)}),R(n)}function R(n){h&&w.includes(n)&&(d.style.colorScheme=n)}function c(){return window.matchMedia("(prefers-color-scheme: dark)").matches?"dark":"light"}if(u)p(u);else try{let n=localStorage.getItem(i)||s,y=l&&n==="system"?c():n;p(y)}catch(n){}})("class","theme","system",null,["light","dark"],null,true,true)
  </script>
  <div
    class="_container_301bee"
  >
    <div
      class="chakra-stack css-fybcnu"
    >
      <form
        role="form"
      >
        <div
          class="chakra-stack css-1igwmid"
        >
          <input
            class="chakra-input css-1hens3x"
            data-testid="name-search-input"
            name="name"
            placeholder="Nombre"
            type="text"
          />
          <button
            class="chakra-button css-4hnymn"
            data-testid="submit-button"
            type="submit"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 24 24"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M10.5 3.75a6.75 6.75 0 1 0 0 13.5 6.75 6.75 0 0 0 0-13.5ZM2.25 10.5a8.25 8.25 0 1 1 14.59 5.28l4.69 4.69a.75.75 0 1 1-1.06 1.06l-4.69-4.69A8.25 8.25 0 0 1 2.25 10.5Z"
                fill-rule="evenodd"
              />
            </svg>
            Buscar
          </button>
        </div>
      </form>
      <div
        class="chakra-stack css-fybcnu"
      >
        <div
          class="css-1uxpq54"
        >
          <table
            class="chakra-table__root css-a2tetu"
          >
            <thead
              class="chakra-table__header css-16u7ocv"
            >
              <tr
                class="chakra-table__row css-11ot24d"
              >
                <th
                  class="chakra-table__columnHeader css-kim7sg"
                >
                  Nombre
                </th>
                <th
                  class="chakra-table__columnHeader css-kim7sg"
                >
                  Descripción
                </th>
                <th
                  class="chakra-table__columnHeader css-kim7sg"
                >
                  Acciones
                </th>
              </tr>
            </thead>
            <tbody
              class="chakra-table__body css-0"
            >
              <tr
                class="chakra-table__row css-11ot24d"
              >
                <td
                  class="chakra-table__cell css-1luypx0"
                >
                  Post 1
                </td>
                <td
                  class="chakra-table__cell css-1luypx0"
                >
                  Description 1
                </td>
                <td
                  class="chakra-table__cell css-1luypx0"
                >
                  <button
                    class="chakra-button css-1uynepj"
                    data-testid="delete-post=button"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 20 20"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
                        fill-rule="evenodd"
                      />
                    </svg>
                    Eliminar
                  </button>
                </td>
              </tr>
              <tr
                class="chakra-table__row css-11ot24d"
              >
                <td
                  class="chakra-table__cell css-1luypx0"
                >
                  Post 2
                </td>
                <td
                  class="chakra-table__cell css-1luypx0"
                >
                  Description 2
                </td>
                <td
                  class="chakra-table__cell css-1luypx0"
                >
                  <button
                    class="chakra-button css-1uynepj"
                    data-testid="delete-post=button"
                    type="button"
                  >
                    <svg
                      aria-hidden="true"
                      fill="currentColor"
                      height="1em"
                      stroke="currentColor"
                      stroke-width="0"
                      viewBox="0 0 20 20"
                      width="1em"
                      xmlns="http://www.w3.org/2000/svg"
                    >
                      <path
                        clip-rule="evenodd"
                        d="M9 2a1 1 0 00-.894.553L7.382 4H4a1 1 0 000 2v10a2 2 0 002 2h8a2 2 0 002-2V6a1 1 0 100-2h-3.382l-.724-1.447A1 1 0 0011 2H9zM7 8a1 1 0 012 0v6a1 1 0 11-2 0V8zm5-1a1 1 0 00-1 1v6a1 1 0 102 0V8a1 1 0 00-1-1z"
                        fill-rule="evenodd"
                      />
                    </svg>
                    Eliminar
                  </button>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
        <div
          class="chakra-group css-6l0f26"
        >
          <p
            class="css-0"
            data-first=""
            data-group-item=""
            data-testid="page-detail"
            style="--group-count: 3; --group-index: 0;"
          >
            Página 1 de 1
          </p>
          <button
            class="chakra-button css-4xchnm"
            data-between=""
            data-group-item=""
            data-testid="previous-button"
            disabled=""
            style="--group-count: 3; --group-index: 1;"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 20 20"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M12.707 5.293a1 1 0 010 1.414L9.414 10l3.293 3.293a1 1 0 01-1.414 1.414l-4-4a1 1 0 010-1.414l4-4a1 1 0 011.414 0z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
          <button
            class="chakra-button css-4xchnm"
            data-group-item=""
            data-last=""
            data-testid="next-button"
            disabled=""
            style="--group-count: 3; --group-index: 2;"
            type="button"
          >
            <svg
              aria-hidden="true"
              fill="currentColor"
              height="1em"
              stroke="currentColor"
              stroke-width="0"
              viewBox="0 0 20 20"
              width="1em"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                clip-rule="evenodd"
                d="M7.293 14.707a1 1 0 010-1.414L10.586 10 7.293 6.707a1 1 0 011.414-1.414l4 4a1 1 0 010 1.414l-4 4a1 1 0 01-1.414 0z"
                fill-rule="evenodd"
              />
            </svg>
          </button>
        </div>
      </div>
      <form
        role="form"
      >
        <div
          class="chakra-stack css-1igwmid"
        >
          <input
            class="chakra-input css-1hens3x"
            data-testid="name-input"
            name="name"
            placeholder="Nombre"
            type="text"
          />
          <input
            class="chakra-input css-1hens3x"
            data-testid="description-input"
            name="description"
            placeholder="Descripción"
            type="text"
          />
          <button
            class="chakra-button css-4hnymn"
            data-testid="submit-button"
            type="submit"
          >
            Crear
          </button>
        </div>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
